src/bin/ion/commands/beta/generate/templates/rust/struct.templ

```diff
@@ -6,68 +6,75 @@ use crate::ion_data_model::{{ import.name | snake }}::{{ import.name | upper_cam
 #[derive(Debug, Clone, Default)]
 pub struct {{ target_kind_name }} {
 {% for field in fields -%}
-         {{ field.name | indent(first = true) }}: {{ field.value }},
+         {{ field.name | snake | indent(first = true) }}: {{ field.value }},
 {% endfor %}
 }
 
 impl {{ target_kind_name }} {
-    pub fn new({% for field in fields -%}{{ field.name }}: {{ field.value }},{% endfor %}) -> Self {
+    pub fn new({% for field in fields -%}{{ field.name | snake }}: {{ field.value }},{% endfor %}) -> Self {
         Self {
             {% for field in fields -%}
-            {{ field.name }},
+            {{ field.name | snake }},
             {% endfor %}
         }
     }
 
 
-    {% for field in fields -%}pub fn {{ field.name }}(&self) -> &{{ field.value }} {
-        &self.{{ field.name }}
+    {% for field in fields -%}pub fn {{ field.name | snake }}(&self) -> &{{ field.value }} {
+        &self.{{ field.name | snake }}
     }
     {% endfor %}
 
 
     pub fn read_from(reader: &mut Reader) -> IonResult<Self> {
+        {% if is_root_type %}
         reader.next()?;
+        {% endif %}
+        let mut abstract_data_type = {{ target_kind_name }}::default();
         {% if abstract_data_type == "Value"%}
-            let mut abstract_data_type = {{ target_kind_name }}::default();
             abstract_data_type.value = {% if target_kind_name | is_built_in_type == false %}
                                             {{ target_kind_name }}::read_from(reader)?;
                                         {% else %}
                                             reader.read_{{ target_kind_name }}()?;
                                         {% endif %}
-        {% else %}
-            {% if abstract_data_type == "Struct"%}reader.step_in()?;{% endif %}
-            let mut abstract_data_type = {{ target_kind_name }}::default();
+        {% elif abstract_data_type is object and abstract_data_type | get(key="Structure") %}
```

Comment:
`abstract_data_type` has following signature,
```
pub enum AbstractDataType {
    // A scalar value (e.g. a string or integer or user defined type)
    // e.g. Given below ISL,
    // ```
    // type::{
    //   name: value_type,
    //   type: int
    // }
    // ```
    // Corresponding abstract type in Rust would look like following:
    // ```
    // struct ValueType {
    //    value: i64
    // }
    // ```
    Value,
    // A series of zero or more values whose type is described by the nested `String` (e.g. a list)
    // e.g. Given below ISL,
    // ```
    // type::{
    //   name: sequence_type,
    //   element: int
    // }
    // ```
    // Corresponding abstract type in Rust would look like following:
    // ```
    // struct SequenceType {
    //    value: Vec<i64>
    // }
    // ```
    Sequence(String),
    // A collection of field name/value pairs (e.g. a map)
    // the nested boolean represents whether the struct has closed fields or not
    // e.g. Given below ISL,
    // ```
    // type::{
    //   name: struct_type,
    //   fields: {
    //      a: int,
    //      b: string,
    //   }
    // }
    // ```
    // Corresponding abstract type in Rust would look like following:
    // ```
    // struct StructType {
    //    a: i64,
    //    b: String,
    // }
    // ```
    Structure(bool),
}
```
This when serialized using `serde` , gives `{Structure:  BOOLEAN_VALUE}`. Hence we have to check which type of abstract data type it is by getting the key value.

Reply:
Okay. I thought the name `struct.templ` meant that it was for `Structure`s.

I think it's worth having separate templates for the different variants here. If that's not possible with Tera, then we should have the `{% if abstract_data_type == "Value"%}`, etc. at the top level of the file so that we're not interleaving the value, sequence, and structure templates.

Reply:
All of these data types do generate a Rust struct, which is why I used the same template.
I can have different templates for all these kinds but they would look almost alike except from some of the read-write statements.

So when we have `AbstractDataType::Sequence`, it also generates a struct but as following:
```rust
struct SequenceType {
    value: Vec<TYPE>
}
```
Same for `ASbstractDataType::Value`,
```rust
struct ValueType {
     value: i64
}
```

Reply:
I understand that they all generate rust structs.

However, reading a templating language is already a little difficult because you have to (at least symbolically) evaluate all of the templating expressions in your mind as you're trying to read it. I'm suggesting that the templates be organized by their abstract data type to make it easier to read and understand them.

Also, is there any particular reason we're creating wrapper types for primitives and `Vec`? And if we are creating wrapper types for them, why named fields instead of `struct ValueType(i64)` or `struct MyTypeVec(Vec<MyType>)`?

Reply:
I can definitely change it to be `struct ValueType(i64)` or `struct MyTypeVec(Vec<MyType>)` but I used wrappers so that we can add `read_from` and `write_to` APIs for these primitives.

Reply:
https://github.com/amazon-ion/ion-cli/issues/93

Reply:
So, that is why I mentioned `WriteAsIon` in one of my other comments. `WriteAsIon` is already implemented for primitive types and for `Vec<T: WriteAsIon>`, and if we don't have an equivalent `ReadFromIon` or similar, then we need to create an issue for that in `ion-rust`.

On the other hand, I think there could also be reason to use wrapper if we want to add additional validation constraints, such as `type::{ name: small_int, type: int, valid_values: range::[0, 9] }`. In that case having `struct SmallInt(u8)` would actually allow us to add the range check to the type. That being said, I can't remember if we decided whether we want to ever have additional validations like that.

Reply:
We do eventually want to add this type validation but not with the initial version of code generator.